node/src/types/block.rs

```diff
@@ -2978,54 +2997,106 @@ mod tests {
         );
     }
 
-    struct TestBlock(Block, TestRng);
+    // Utility struct that can be turned into an iterator that generates
+    // continuous and descending blocks (i.e. blocks that have consecutive height
+    // and parent hashes are correctly set). The height of the first block
+    // in a series is choosen randomly.
+    //
+    // Additionally, this struct allows to generate switch blocks at a specific location in the
+    // chain, for example: Setting `switch_block_indices` to [1; 3] and generating 5 blocks will
+    // cause the 2nd and 4th blocks to be switch blocks.
+    struct TestBlockSpec {
+        block: Block,
+        rng: TestRng,
+        switch_block_indices: Option<Vec<u64>>,
+    }
 
-    impl TestBlock {
-        fn new(test_rng: TestRng) -> Self {
+    impl TestBlockSpec {
+        fn new(test_rng: TestRng, switch_block_indices: Option<Vec<u64>>) -> Self {
             let mut rng = test_rng;
-            let init = Block::random(&mut rng);
-            Self(init, rng)
+            let block = Block::random(&mut rng);
+            Self {
+                block,
+                rng,
+                switch_block_indices,
+            }
         }
 
         fn into_iter(self) -> TestBlockIterator {
-            TestBlockIterator(self.0, self.1)
+            let block_height = self.block.height();
+            TestBlockIterator {
+                block: self.block,
+                rng: self.rng,
+                switch_block_indices: self.switch_block_indices.map(|switch_block_indices| {
+                    switch_block_indices
+                        .iter()
+                        .map(|index| index + block_height)
+                        .collect()
+                }),
+            }
         }
     }
 
     const NEVER_SWITCH_HASHING: EraId = EraId::new(u64::MAX);
 
-    struct TestBlockIterator(Block, TestRng);
+    struct TestBlockIterator {
+        block: Block,
+        rng: TestRng,
+        switch_block_indices: Option<Vec<u64>>,
+    }
 
     impl Iterator for TestBlockIterator {
         type Item = Block;
 
         fn next(&mut self) -> Option<Self::Item> {
+            let (is_switch_block, validators) = match &self.switch_block_indices {
+                Some(switch_block_indices)
+                    if switch_block_indices.contains(&self.block.height()) =>
```

Comment:
Shouldn't we use the _next block_ here rather than the previous one?

Reply:
No, I think the logic is correct. The switch blocks are created at the correct spots as proven in [this test](https://github.com/casper-network/casper-node/pull/3125/files#diff-f8f1402a94ae96f1e79dfff1d2af53d86a2114c73096a211873dbbb4fe4226e9R3117).